### PR TITLE
Introduce an API for variant dependency substitution

### LINKED ISSUE #13204: Dependency Substitution not working on platform project dependencies.
Created a sample proejct for the issue at [https://github.com/udayshnk/gradle-platform-dependency-substitution](https://github.com/udayshnk/gradle-platform-dependency-substitution).

Its a multi module project with subprojects named **foo-v1**, **foo-v2**, **baz-v1**, **baz-v2** and **bar**.
**bar** is depending on **foo-v1** using platform project dependency and also depending on **baz-v1** as a regular project dependency.

    dependencies {
        implementation(platform(project(":foo-v1")))
        implementation(project(":baz-v1"))
        implementation("com.google.guava:guava")
        implementation("com.fasterxml.jackson.core:jackson-databind")
    }

Dependency substitution is declared in parent build.gradle.kts to substiture **foo-v1**,**baz-v1** with **foo-v2**,**baz-v2** respectively as below

    subprojects {
       configurations {
           all {
               resolutionStrategy {
                   dependencySubstitution {
                       substitute(project(":foo-v1")).apply {
                           with(project(":foo-v2"))
                       }
                       substitute(project(":baz-v1")).apply {
                           with(project(":baz-v2"))
                       }
                   }
               }
           }
       }
    }

Below is the dependency graph for runtime configuration of project **bar**

    +--- project :foo-v1
    |    +--- com.fasterxml.jackson:jackson-bom:2.9.7
    |    |    +--- com.fasterxml.jackson.core:jackson-databind:2.9.7 (c)
    |    |    +--- com.fasterxml.jackson.core:jackson-annotations:2.9.0 (c)
    |    |    \--- com.fasterxml.jackson.core:jackson-core:2.9.7 (c)
    |    \--- com.google.guava:guava:28.0-jre
    |         +--- com.google.guava:failureaccess:1.0.1
    |         +--- com.google.guava:listenablefuture:9999.0-empty-to-avoid-conflict-with-guava
    |         +--- com.google.code.findbugs:jsr305:3.0.2
    |         +--- org.checkerframework:checker-qual:2.8.1
    |         +--- com.google.errorprone:error_prone_annotations:2.3.2
    |         +--- com.google.j2objc:j2objc-annotations:1.3
    |         \--- org.codehaus.mojo:animal-sniffer-annotations:1.17
    +--- project :baz-v1 -> project :baz-v2
    |    \--- commons-io:commons-io:2.6
    +--- com.google.guava:guava -> 28.0-jre (*)
    \--- com.fasterxml.jackson.core:jackson-databind -> 2.9.7
         +--- com.fasterxml.jackson.core:jackson-annotations:2.9.0
         \--- com.fasterxml.jackson.core:jackson-core:2.9.7

From the dependenecy graph it can be seen that dependency substitution is working fine for regular dependency like **baz**, whereas for platform dependency like **foo** substitution is not working.

Note: I have tried to debug the gradle code to figure out the issue. I have observed that DefaultProjectComponentSelector object which is constructed by parsing build script from dependencySubstituion has missing attribute **org.gradle.category=platform** for platform project dependencies.

### PR DESCRIPTION
**Pull request by melix**
This commit introduces a new API to deal with _variant-aware_ dependency
substitution. Before this change, it wasn't possible to substitute a
dependency on a platform, for example. This change allows configuring
the attributes of a dependency, both on the "source" side (what dependency
we need to substitute) and the "target" side (what dependency is going to
be used instead).

Some DSL short-hand notations are added for common use cases.

Fixes #13204

### CODE REVIEW

**Comment by melix** [comment]
@bot-gradle test RFM

**Comment by melix** [comment]
@ljacomet would appreciate an early review wrt the DSL

**Comment by melix** [comment]
@bot-gradle test RFM

**Comment by melix** [comment]
@bot-gradle test RFM

**Comment by melix** [comment]
@bot-gradle test RFM

File: subprojects/core-api/src/main/java/org/gradle/api/artifacts/VariantSelectionDetails.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+    /**
+     * Selects the library variant of a component
+     */
+    void library();
+
+    /**
+     * Replaces the provided selector attributes with the attributes configured
+     * in the configuration action.
+     * @param configurationAction the configurationa action
```
Comment (ljacomet) [inline subprojects/core-api/src/main/java/org/gradle/api/artifacts/VariantSelectionDetails.java:50]: Typo: trailing `a` in configuration

File: subprojects/core-api/src/main/java/org/gradle/api/artifacts/DependencySubstitutions.java [hunk 2]
```diff
@@ -71,6 +72,25 @@
...
 
+    /**
+     * Transforms the supplied selector into a specific variant selector.
+     * @param selector the origin selector
+     * @param detailsAction the variant selection details configuration
+     *
+     * @since 6.6
+     */
+    @Incubating
+    ComponentSelector variant(ComponentSelector selector, Action<? super VariantSelectionDetails> detailsAction);
```
Comment (ljacomet) [inline subprojects/core-api/src/main/java/org/gradle/api/artifacts/DependencySubstitutions.java:83]: I am not sure about the best place in the javadoc, but we should clearly state that the attributes specified here need only be the ones _different_ from those of the resolution context.
Maye we can simply reference the documentation for adding attributes on dependency declarations. The applicability is the same.

**Pull request merged**